offset_manager.go

```diff
@@ -74,52 +86,292 @@ func (om *offsetManager) ManagePartition(topic string, partition int32) (Partiti
+	request := new(OffsetFetchRequest)
+	request.Version = 1
+	request.ConsumerGroup = om.group
+	request.AddPartition(topic, partition)
 
-	bom := om.boms[broker]
-	if bom == nil {
-		bom = om.newBrokerOffsetManager(broker)
-		om.boms[broker] = bom
+	response, err := broker.FetchOffset(request)
+	if err != nil {
+		if retries <= 0 {
+			return 0, "", err
+		}
+		om.releaseCoordinator(broker)
+		return om.fetchInitialOffset(topic, partition, retries-1)
 	}
 
-	bom.refs++
+	block := response.GetBlock(topic, partition)
+	if block == nil {
+		return 0, "", ErrIncompleteResponse
+	}
 
-	return bom
+	switch block.Err {
+	case ErrNoError:
+		return block.Offset, block.Metadata, nil
+	case ErrNotCoordinatorForConsumer:
+		if retries <= 0 {
+			return 0, "", block.Err
+		}
+		om.releaseCoordinator(broker)
+		return om.fetchInitialOffset(topic, partition, retries-1)
+	case ErrOffsetsLoadInProgress:
+		if retries <= 0 {
+			return 0, "", block.Err
+		}
+		select {
+		case <-om.closing:
+			return 0, "", block.Err
+		case <-time.After(om.conf.Metadata.Retry.Backoff):
+		}
+		return om.fetchInitialOffset(topic, partition, retries-1)
+	default:
+		return 0, "", block.Err
+	}
 }
 
-func (om *offsetManager) unrefBrokerOffsetManager(bom *brokerOffsetManager) {
-	om.lock.Lock()
-	defer om.lock.Unlock()
+func (om *offsetManager) coordinator() (*Broker, error) {
+	om.brokerMu.RLock()
+	broker := om.broker
+	om.brokerMu.RUnlock()
+
+	if broker != nil {
+		return broker, nil
+	}
 
-	bom.refs--
+	om.brokerMu.Lock()
+	defer om.brokerMu.Unlock()
 
-	if bom.refs == 0 {
-		close(bom.updateSubscriptions)
-		if om.boms[bom.broker] == bom {
-			delete(om.boms, bom.broker)
+	if broker := om.broker; broker != nil {
+		return broker, nil
+	}
+
+	if err := om.client.RefreshCoordinator(om.group); err != nil {
+		return nil, err
+	}
+
+	broker, err := om.client.Coordinator(om.group)
+	if err != nil {
+		return nil, err
+	}
+
+	om.broker = broker
+	return broker, nil
+}
+
+func (om *offsetManager) releaseCoordinator(b *Broker) {
+	om.brokerMu.Lock()
+	if om.broker == b {
+		om.broker = nil
+	}
+	om.brokerMu.Unlock()
+}
+
+func (om *offsetManager) mainLoop() {
+	defer om.ticker.Stop()
+	defer close(om.closed)
+
+	for {
+		select {
+		case <-om.ticker.C:
+			om.flushToBroker()
+			om.releasePOMs(false)
+		case <-om.closing:
+			return
 		}
 	}
 }
 
-func (om *offsetManager) abandonBroker(bom *brokerOffsetManager) {
-	om.lock.Lock()
-	defer om.lock.Unlock()
+func (om *offsetManager) flushToBroker() (success bool) {
+	request := om.constructRequest()
+	if request == nil {
+		return true
+	}
+
+	broker, err := om.coordinator()
+	if err != nil {
+		om.handleError(err)
+		return false
+	}
+
+	response, err := broker.CommitOffset(request)
+	if err != nil {
+		om.handleError(err)
+		om.releaseCoordinator(broker)
+		_ = broker.Close()
+		return false
+	}
+
+	success = true
+
+	om.pomsMu.Lock()
+	defer om.pomsMu.Unlock()
 
-	delete(om.boms, bom.broker)
+	for _, topicManagers := range om.poms {
```

Comment:
Just for organization I'd prefer `handleResponse` or something as a separate method to mirror `constructRequest`